Some fixes for wxQt

AliKet opened the pull request:
Replaces #24326

The last two commits are the most questionable on this PR. And AFAICT they works perfectly well and I didn't notice any regression so far. But please let me know if you see anything wrong with them. TIA!

Commit:
wxQt: Move DoGetBestSize() function from wxWindow to wxControl

DoGetBestSize() doesn't have to be in wxWindow itself, and it makes more sense
to put it in wxControl instead. Also simplify its implementation a little bit.

Commit:
wxQt: Fix wxWindow::Show() implementation

This function generates wxEVT_SHOW event (from QtHandleShowEvent()) and not
wxEVT_SIZE event. Also, make it return true, like the other ports do, if the
function is called before the underlying window is created.

Commit:
wxQt: Update comment in wxWindowQt::PostCreation()

The comment is not entirely correct. because in fact, we can create a window
initially hidden under wxQt. For example, if setVisible(true) is not called for
the created widget, it will remain hidden.

Commit:
wxQt: Fix the initial client size of an initially hidden window.

Since the TLWs are initially hidden, we need to postpone calling QWidget::geometry()
in wxWindow::DoGetClientSize() until the Qt widget is actually shown.

Fixes wxAUI broken layout under wxQt.

Commit:
wxQt: Fix wx{Window,Client}DC::GetSize() returning invalid sizes

These DCs use QPicture which records and replays QPainter commands. Calling GetSize()
(which calls wxQtDCImpl::DoGetSize()) on these DCs returns (0, 0) if QPicture doesn't
have any data. So just do what wxMSW and wxGTK do by returning the correct size based
on the type of the DC.

Commit:
Use GetBestSize() instead of GetClientSize() when adding a pane to wxAuiManager

It doesn't seem like it matters which function is used under the other ports.
But under wxQt, GetClientSize() does not return the expected "best size" like
the other ports do.

Also fix wxAuiToolBar::GetBestSize() under wxMSW.

vadz commented on include/wx/aui/auibar.h:
```diff
@@ -627,6 +627,12 @@ class WXDLLIMPEXP_AUI wxAuiToolBar : public wxControl
+    // returns DEFAULT_ITEM_HEIGHT which is not suitable as the default height/width for
+    // a horizontal/vertical toolbar. Notice that the toolbar will be resized correctly
+    // when Realize() is called.
+    virtual wxSize DoGetBestSize() const override { return wxDefaultSize; }
```
It seems very suspicious to return `wxDefaultSize` from here. I'm not sure it breaks something, but I'd be surprised if it didn't.

Couldn't we return `m_sizer->GetMinSize()` here instead? If not, let's at least return `wxSize(1,1)` or something like this, i.e. a valid size, even if a wrong one.

vadz replied:
Sorry, but returning current size (i.e. `GetClientSize()`) as best size is definitely not a good idea because this means the window can never become smaller than its current size. If `GetMinSize()` has a valid value, let's return it instead. If not, returning `wxSize(1, 1)` at least shouldn't break anything (although could result in the window being too small, of course).

vadz commented on src/aui/framemanager.cpp:
```diff
@@ -799,17 +799,7 @@ bool wxAuiManager::AddPane(wxWindow* window, const wxAuiPaneInfo& paneInfo)
-            pinfo.best_size = pinfo.window->GetBestSize();
-        }
-#endif // wxUSE_TOOLBAR
+        pinfo.best_size = pinfo.window->GetBestSize();
```
I'd like to get rid of the toolbar hack (which I don't really understand), but this is a change in behaviour as the pane will use its minimal, and not current, size now.

I don't know if it's really important, but neither am I sure that it isn't :-(

vadz reviewed (commented):
Sorry, I've somehow completely forgot to look at this one.

I have no comments about all the previous commits, but the last one is indeed questionable. If you'd like, we could merge all the other ones already and deal with this one later.

Thanks!

vadz commented:
Thanks for the update, I'll merge it soon (forgot to do it during the last round of merges, sorry).

AliKet commented:
> Thanks for the update, I'll merge it soon (forgot to do it during the last round of merges, sorry).

This is fine as I have 2 bug fixes affecting `wxAUI` that I want be included in this PR. TIA!

Commit:
wxQt: Simplify wxWindowQt::QtReparent()

No real changes.

Commit:
wxQt: Fix broken mouse capture that was not handled properly

Reproducible in wxAUI sample: when changing a tab in wxAuiNotebook,
it prompts us to confirm with a dialog box which we cannot interact
with because the tab grabs the mouse.

Notice also that we don't need to grab/release the mouse explicitly
as it is done implicitly for us by Qt:

"It is seldom necessary to grab the mouse when using Qt, as Qt grabs
 and releases it sensibly. In particular, Qt grabs the mouse when a
 mouse button is pressed and keeps it until the last button is released."

AliKet commented:
@vadz `wxQt` and `wxGTK2` CI need to be restarted , TIA!
